Fixes for issues found with META6

jonathanstowe opened the pull request:
* Tighten up the _unmarshal(%, Mu) candidate to not match Positional|Associative target
* panic on Associate data for Positional target
* Also add test of META6 in GH action as a canary

Fixes #7

jonathanstowe committed:
Fixes for issues found with META6

* Tighten up the _unmarshal(%, Mu) candidate to not match
  Positional|Associative target
* panic on Associate data for Positional target
* Also add test of META6 in GH action as a canary

Fixes #7

jonathanstowe committed:
Temporarily remove the META6 as current test data will fail

jonathanstowe commented:
Not in a position to diagnose the fail of App::Prove6 on Windows.

jonathanstowe committed:
Simplify the where of the (%, Mu) candidate

The order of the smartmatch is non-obvious, but it doesn't work the
other way round.

vrurg commented:
Turns out I was commenting on an outdated commit of this PR. I think the comment https://github.com/raku-community-modules/JSON-Unmarshal/commit/61627a693f3e21db1c9c81bdf30af6256a7a92cf#r122446733 is relevant and worth referencing to.

vrurg commented on lib/JSON/Unmarshal.rakumod:
```diff
@@ -248,7 +248,11 @@ multi _unmarshal(Any:D $json, Bool) {
 }
 
-multi _unmarshal(%json, Mu $obj is raw) {
+multi _unmarshal(%json, @x) {
```
There going to be a problem about this for classes trying to implement the Positional role but remain serializable otherwise (a pseudo-code example):

```raku
class MyPos does Positional {
    has @!values;
    has Str:D $.source; # Where we got our values from
    has UserData $.something-else;
    method AT-POS($idx) { self!is-ok-position ?? @!values[$idx] !! self!throw-bad-pos }
    ...
}
```

jonathanstowe replied:
Yes. But unfortunately without this "guard"  it will blunder into another candidate where it will blow up in some uncaught fashion.

I think another test is called for :smile_cat:

jonathanstowe replied:
It appears that it would have blown up in the `unmarshal(Str:D $json, Positional $obj, *%c) `  even without that candidate with something like:

```
    my class TestClass does Positional {
        has Str $.string;
    }

    my $string-attribute = '{ "string" : "test value" }';

    my $obj;

    $obj = unmarshal($string-attribute, TestClass);

    is $obj.string, 'test value', "got the attribute back";
```

On a typed attribute of another class it does fail in that candidate.  I think both can be solved with a subset which is like the reverse of the where clause you have.  Let's see.

jonathanstowe replied:
See the latest commit which deals with both of those cases.

vrurg commented:
Eventually, I came down to a conclusion that there is no satisfiable solution for the problem. Let me sum up a bit.

The problematic candidate is for cases where a key/value serialization must be deserialized into object attributes. This means there must be at least one public attribute on the object for the current implementation.

We must not care about additional constraints on the target attribute. `when` purpose is to constraint attribute values, not to determine its type. Even though `when Positional | Associative` is a type-like thing, attribute's type remains `Mu`. We cannot deserialize a hashy JSON into pure Mu as in an object.

It all winds down to the proposal in the above mentioned comment of mine. Except that in a hurry I used `grep` where `first` would do better. 

My only doubt about it is that it is potentially slow as `when` with a code block is non-optimizable by its nature.

I about to push my proposed commit into this branch in a minute. Feel free to discard it if anything.

vrurg committed:
Make object deserialization candidate do the right thing

Thow it would be slower than any type matching.

jonathanstowe committed:
Deal better with types that implement Positional

A type that implemented Positional but has accessors (so presumably is
amenable to a pairwise constructor,) wouldn't have worked before.

This may not work for a Positional which has a constructor which takes
a list of things but by that point I'd expect the implementor to provide
their own BUILD or custom constructor.

jonathanstowe commented:
If everyone is happy I'll merge this, but could someone make the release as soon as practical as I will need to release a new `JSON::Class` to bump the dependency version, and revisions to the tests in `META6`.

lizmat commented:
Release 0.14 in the ecosystem as of now.
